src/datastructure/Map.cpp

```diff
@@ -199,87 +199,15 @@ bool Map::hasKeyframeImages() const
-    copyPointCloud->setDescriptorType(m_pointCloud->getDescriptorType());
-    std::vector<SRef<CloudPoint>> srefPoints;
-    m_pointCloud->getAllPoints(srefPoints);
-    std::vector<CloudPoint> points;
-    for (const auto & pt: srefPoints) {
-        points.push_back(*pt);
-    }
-    copyPointCloud->addPoints(points, false);
-
-    SRef<KeyframeCollection> copyKeyframeCollection = mapWithoutKeyframeImages->getConstKeyframeCollection();
-    SRef<KeyframeRetrieval> copyKeyframeRetrieval = mapWithoutKeyframeImages->getConstKeyframeRetrieval();
-    copyKeyframeCollection->setDescriptorType(m_keyframeCollection->getDescriptorType());
-    std::vector<SRef<Keyframe>> keyframesWithoutImages;
-    m_keyframeCollection->getAllKeyframesWithoutImages(keyframesWithoutImages);
-    for (const auto & kf: keyframesWithoutImages) {
-        copyKeyframeCollection->addKeyframe(kf, false);
-        uint32_t kfID = kf->getId();
-        BoWFeature bowDesc;
-        m_keyframeRetrieval->getBoWFeature(kfID, bowDesc);
-        BoWLevelFeature bowLevelDesc;
-        m_keyframeRetrieval->getBoWLevelFeature(kfID, bowLevelDesc);
-        copyKeyframeRetrieval->addDescriptor(kfID, bowDesc, bowLevelDesc);
-    }
-
-    SRef<CovisibilityGraph> copyCovisibilityGraph = mapWithoutKeyframeImages->getConstCovisibilityGraph();
-    std::set<uint32_t> nodesID;
-    m_covisibilityGraph->getAllNodes(nodesID);
-    for (const auto & nID: nodesID) {
-        std::vector<uint32_t> neighbors;
-        m_covisibilityGraph->getNeighbors(nID, 0, neighbors);
-        for (const auto & nb: neighbors) {
-            float weight;
-            m_covisibilityGraph->getEdge(nID, nb, weight);
-            copyCovisibilityGraph->increaseEdge(nID, nb, weight);
-        }
-    }
-
-    SRef<CameraParametersCollection> copyCameraParametersCollection = mapWithoutKeyframeImages->getConstCameraParametersCollection();
-    std::vector<SRef<CameraParameters>> cameraParameters;
-    m_cameraParametersCollection->getAllCameraParameters(cameraParameters);
-    for (const auto & camParam: cameraParameters) {
-        CameraParameters copyCamParam;
-        copyCamParam.name = camParam->name;
-        copyCamParam.id = camParam->id;
-        copyCamParam.type = camParam->type;
-        copyCamParam.resolution = camParam->resolution;
-        copyCamParam.intrinsic = camParam->intrinsic;
-        copyCamParam.distortion = camParam->distortion;
-        copyCamParam.extrinsics = camParam->extrinsics;
-        copyCameraParametersCollection->addCameraParameters(copyCamParam, false);
-    }
-
-    mapWithoutKeyframeImages->setVersion(m_version);
-    mapWithoutKeyframeImages->setDescriptorType(m_descriptorType);
-    mapWithoutKeyframeImages->setGlobalDescriptorType(m_globalDescriptorType);
     if (m_embedKeyframeImages) {
-        mapWithoutKeyframeImages->embedKeyframeImages();
+        std::vector<SRef<Keyframe>> keyframes;
+        m_keyframeCollection->getAllKeyframes(keyframes);
```

Comment:
`m_keyframeCollection->nextSerializationWithoutKeyframeImages()`?

Reply:
Since this flag is deep in the Map datastructure (Map->KeyframCollection->Frame) and each object is likely shared all over the place.

So there also a risk that, if some frames are serialized somehow  via a different path, they will reset their serialization flag, leading to them being unwillingly serialized while serializing the Map.

Maybe a way to lower the risk would be to make Map `friend` of KeyframeCollection (if you decide to implement the method as suggested above), and KeyframeCollection `friend` of Frame.

And make `nextSerializationWithoutKeyframeImages()` for `KeyframeCollection` and `Frame` private (or maybe protected, because of the inheritance between Frame and Keyframe, I  don't know).

This would force to enable this feature only through the Map object.